storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +251,225 @@ func (s *commitLogSource) Read(
-						defer stream.Finalize()
-					}
-
-					iter := multiReaderIteratorPool.Get()
-					iter.Reset(readers)
-
-					var err error
-					enc := encoderPool.Get()
-					enc.Reset(start, blopts.DatabaseBlockAllocSize())
-					for iter.Next() {
-						dp, unit, annotation := iter.Current()
-						encodeErr := enc.Encode(dp, unit, annotation)
-						if encodeErr != nil {
-							err = encodeErr
-							errs++
-							break
-						}
-					}
-					if iterErr := iter.Err(); iterErr != nil {
-						if err == nil {
-							err = iter.Err()
-						}
-						errs++
-					}
-
-					iter.Close()
-					for i := range unmergedBlock {
-						unmergedBlock[i].enc.Close()
-					}
-
-					if err != nil {
-						continue
-					}
-
-					block.Reset(start, enc.Discard())
-				}
-				blocks.AddBlock(block)
+		mergeSemaphore <- true
+		wg.Add(1)
+		go func(shard int, unmergedShard encodersAndRanges) {
+			var shardResult result.ShardResult
+			shardResult, shardEmptyErrs[shard], shardErrs[shard] = s.mergeShard(
+				unmergedShard, blocksPool, multiReaderIteratorPool, encoderPool, blopts)
+			if len(shardResult.AllSeries()) > 0 {
+				// Prevent race conditions while updating bootstrapResult from multiple go-routines
+				bootstrapResultLock.Lock()
+				// Shard is a slice index so conversion to uint32 is safe
+				bootstrapResult.Add(uint32(shard), shardResult, nil)
+				bootstrapResultLock.Unlock()
+			}
+			<-mergeSemaphore
+			wg.Done()
+		}(shard, unmergedShard)
+	}
+
+	// Wait for all merge goroutines to complete
+	wg.Wait()
+	s.logMergeShardsOutcome(shardErrs, shardEmptyErrs)
+	return bootstrapResult
+}
+
+func (s *commitLogSource) mergeShard(
+	unmergedShard encodersAndRanges,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (shardResult result.ShardResult, numShardEmptyErrs int, numErrs int) {
+	shardResult = result.NewShardResult(len(unmergedShard.encodersBySeries), s.opts.ResultOptions())
+	for _, unmergedBlocks := range unmergedShard.encodersBySeries {
+		seriesBlocks, numSeriesEmptyErrs, numSeriesErrs := s.mergeSeries(
+			unmergedBlocks,
+			blocksPool,
+			multiReaderIteratorPool,
+			encoderPool,
+			blopts,
+		)
+
+		if seriesBlocks.Len() > 0 {
+			shardResult.AddSeries(unmergedBlocks.id, seriesBlocks)
+		}
+
+		numShardEmptyErrs += numSeriesEmptyErrs
+		numErrs += numSeriesErrs
+	}
+	return shardResult, numShardEmptyErrs, numErrs
+}
+
+func (s *commitLogSource) mergeSeries(
+	unmergedBlocks encodersByTime,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (seriesBlocks block.DatabaseSeriesBlocks, numEmptyErrs int, numErrs int) {
+	seriesBlocks = block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
+
+	for startNano, encoders := range unmergedBlocks.encoders {
+		start := startNano.ToTime()
+		block := blocksPool.Get()
```

Comment:
would make lifecycle easier to track if you only Get a `block` once you're sure you're going to need it. e.g.

```
for ... {
  start := .. 
  
  if len(encoders) == 0 {
    numEmptyErrs++
    continue
  }

  if len(encoders) == 1 {
    block := blocksPool.Get()
    block.Reset(start, ...)
    seriesBlocks.AddBlock(block)
    continue
  }
  
  // .. all the []encoder => encoder conversions 
 
  // .. encoder/iter cleanup 
  
  block := blocksPool.Get(..)
  // .... 
}
```

Reply:
good call